### ISSUE TITLE
Change to the release-process: history & tag based, instead of manual sync

### ISSUE THREAD

**Issue by Keenuts** [KhronosGroup/SPIRV-Tools#5089]
Hello all!

notice: this is written from my point of view when handling the release, I might lack vision on what external partners are doing during this process.

### Today

Today's release process involves multiple steps, done by hand. What this is aiming at is to make most of the flow automatic, and unfreeze main while the release process is ongoing.

Flow today:
 1. choosing a commit hash with passing CI (usually main:HEAD)
 2. submit PR to update CHANGES files with changelog (reported version **depends** on it) and new version: vYYYY.X
 3. tag this PR for release vYYYY.X
 4. submit PR to update CHANGES again, changing version to vYYYY.(X+1)-dev (so spirv-tools built at head shows future version) 
 3. sync canary branch to it
 4. run additional testing (each company running their own internal tests)
 6. if testing fails, we fix and start over with a new release version, vYYYY.(X+1)
 8. rebase the stable branch
 9. create tags (vYYYY.X)
 10. copy-paste CHANGES changelog for release into github release form, and click "create release" using this tag
 

 ### Reasons to fail

 - we rely on a hand-made file (CHANGES) for version tracking, and hope it reflects the git history closely enough
 - because of that, we need those 2 PRs: one to change -dev into final release version, and one to set the next -dev version.
 - lot of manual steps

  
### Proposition

- use git history & tags for version tracking (instead of CHANGES file) (https://github.com/KhronosGroup/SPIRV-Tools/pull/5067)
- protect canary & stable branches
- create github workflow to update canary/stable branches depending on tags.
- create github workflow to create release based on tags

Here is a release flow proposition:
 - submit a PR to update the CHANGES file changelog (as we do today)
 - create a release candidate tag (vYYYY.X.rc1) at main:head
 - github workflow runs when this tag is pushed, rebasing canary to match main at this tag
 - we run our tests on this candidate, we find a bug
 - we fix on main
 - create new candidate tag (vYYYY.X.rc2)
 - github workflow updated canary to match this new tag
 - we test again, it passes
 - we tag the release candidate as vYYYY.X
 - github workflow rebase stable to match this tag, and creates a github release (text sourced from the CHANGES file still). 
 
Main idea is to do most tasks in an automatic fashion, and replace manual steps with less error-prone ones.
(mostly git tag, vs careful editing of CHANGES, PR ordering, and manual rebasing)

Any thoughts? 😊

**Comment by Keenuts** [comment]
Today @s-perron mentioned stable/canary were used at a time when SPIRV-Tools was really unstable, but maybe we could only rely on tags, and stop using those at all.

Leaving us with this new process:
 - submit a PR to update the CHANGES file changelog (as we do today)
 - create a release candidate tag (vYYYY.X.rc1) at main:head
 - we run our tests on this candidate, we find a bug
 - we fix on main
 - create new candidate tag (vYYYY.X.rc2)
 - we test again, it passes
 - we tag the release candidate as vYYYY.X
 - github workflow triggers on this tag (without rcX), and creates a github release (text sourced from the CHANGES file still).

**Comment by s-perron** [comment]
I agree with the change in the process. The "stable" branch was meant to be something that was well tested between releases, but now that development on spirv-tools has slowed, it is always the same as the latest release. The release candidate tags are essentially the canary branches. This makes those branches redundant.

**Comment by greg-lunarg** [comment]
> (text sourced from the CHANGES file still)

This is not parsing for me i.e. its not clear to me what this is saying. Can you please clarify?

Relating to CHANGES, when in the process is the CHANGES PR merged?

Otherwise I don't have any issues with this change to the process.

FWIW, I usually do a testing pass even before the canary / rc is posted to try give myself more time to fix any bugs that have crept in over the last couple months.

**Comment by Keenuts** [comment]
The release notes usually list all the changes since the last release.
AFAIK we never wrote the full list of commit msgs, but selected the important ones to write then in the CHANGES file.
So the goal is to keep this "editorialized" message, while having automatic releases:
 - we manually edit the CHANGES file to write the changelog message for this version.
 - the workflow reads it to create the release message.
 The version is not parsed, only the changelog.
 
 As for the version read from tags, yes, it's live. But not with the vYYYY.N.rcX tag format.

**Comment by Keenuts** [comment]
Yes, but now the CHANGES file content is only informative, so it's OK if we have a wrong CHANGES file for a few commits.
(This is already the case since we don't rely on it for version guess)

The main point of focus now it making the rest automatic:
no more manual push/rebase, less branches etc.

**Comment by Keenuts** [comment]
Procedure after review feedback:

- drop the canary and release branch. This was done at a time when SPIRV-Tools was very unstable.

* 1 maintainer updated the CHANGES file to have an "editorialized" CHANGELOG
   (take the git log, filter important commits, write a nice changelog)
   This is submitted as a normal PR (review + merge)
* 1 maintainer create a new release candidate tag `vYYYY.X.rc1` at `main:head`
* all partners are notified of this release candidate. They should test it and report back results (~1 week window delay)
   (mostly through internal communication channel, as not everything can be public)
* The maintainer who pushed the rc1 tag is notified of a failure. The test is fixed at `main:head`.
* 1 maintainer pushes a new tag `vYYYY.X.rc2`
* we go for another round of partner testing
* the maintainer who pushed the `rc2` tag received an "all good"
* the maintainer pushes a new tag at `vYYYY.X.rc2`: `vYYYY.X`
* a github workflow triggers on this tag, parses the `CHANGES` file to generate the changelog message, and publishes a new release.

**Comment by dneto0** [comment]
> AFAIK we never wrote the full list of commit msgs, but selected the important ones to write then in the CHANGES file.
So the goal is to keep this "editorialized" message

Yes, I think the CHANGES file should be a cut down digest of what happened.  If someone really wants all the detail they can read the git log.

The .rc* mechansim seems fine to me. 

My only concern is using GitHub tags for the code to know what version it is.  The advantage of using the CHANGES file as the source of truth for version info is the files know their own "name".
For example, when we import SPIRV-Tools into another repo, those tags don't come along. It's useful to line up self-reported version strings with the release numbers externally.  

For example SPIRV-Tools is imported into the Android Open Source Project, at https://android.googlesource.com/platform/external/shaderc/spirv-tools/  
It has its own set of tags.

**Comment by Keenuts** [comment]
> My only concern is using GitHub tags
> For example, when we import SPIRV-Tools into another repo, those tags don't come along.

when I mention `tags`, I mention git tags, not something related to github. (`git tags`).
So if it's a submodule, tags will be known. Same for clones, as long as depth is not 0.

> For example SPIRV-Tools is imported into the Android Open Source Project
Don't know if repo-tools prunes git tags or not. Do you know that?

**Comment by Keenuts** [comment]
TL;DR:
 - keep the steps mentioned above, with `vYYYY.X.rcY` tags and validation process. 
 - when a `vYYYY.X` tag is pushed, a new branch is created from this tag.
 - a github workflow created a `VERSION` file, with the SPIRV-Tools version written in it.
 - this file is read as fallback when no git information is available.
 
 Ok, got it now.
 Some context first:
  - before we had this CHANGES file to determine the version.
  - this was fine as long as we have no bugfix/merge while testing the release version.
     -> make a PR to update the CHANGES file
     -> create tag
     -> create github release.
  - the problem was when the release process went wrong (bug found). Then the version reported by CHANGES could be out-of-sync with the release. And this required stop merging new PRs during the release week, and being careful around CHANGES content, leaving room for error.
  
So the a change has been made to use only git tags for version tracking.
This is fine when using SPIRV-Tools as a repo. But not when no git information is available, an example are the release zip files.

So we need a solution to:
 - A: simplify the release process, remove this issue around CHANGES file management.
 - B: allow version tracking when there is no git history.
 
 What I'd suggest is the following:
 
 Procedure after review feedback:

* 1 maintainer updated the CHANGES file to have an "editorialized" CHANGELOG
   (take the git log, filter important commits, write a nice changelog)
   This is submitted as a normal PR (review + merge)
* 1 maintainer create a new release candidate tag `vYYYY.X.rc1` at `main:head`
* all partners are notified of this release candidate. They should test it and report back results (~1 week window delay)
   (mostly through internal communication channel, as not everything can be public)
* The maintainer who pushed the rc1 tag is notified of a failure. The test is fixed at `main:head`.
* 1 maintainer pushes a new tag `vYYYY.X.rc2`
* we go for another round of partner testing
* the maintainer who pushed the `rc2` tag received an "all good"
* the maintainer pushes a new tag at `vYYYY.X.rc2`: `vYYYY.X`
* a github workflow triggers on this tag
    - creates a new branch `vYYYY.X`
    - creates a commit on this branch, adding a VERSION file (content is `vYYYY.X`)
    - parses the `CHANGES` file to generate the changelog message
    - publishes a new release using this branch.
    
 This last step allows us to allow merges while preparing a release.
 We also minimize divergence on the release branch, in fact we have only this `VERSION` file added. No bugfix and stuff.
 And we support using the repo on environments with no git history, as long as they use releases.
 @dneto0 would that be OK?

**Comment by Keenuts** [comment]
Hi all,

Here is the GitHub workflow I'm suggesting for this release process:
https://github.com/Keenuts/SPIRV-Tools/blob/release-workflow/.github/workflows/release.yml

@dneto0 : to allow version reporting for users with no git history, I added the option to create a VERSION file.
The difference is the following:
 - it links the version to a hash in this repo
 - it is **not** generated or used by default. Our release process don't use it.
 - can be generated when importing SPIRV-Tools in the user's infra (Example, AOSP could pull a release branch, which have this, or a user could pull the main branch and generate it on the fly)

**Comment by Keenuts** [comment]
Hello all!

After some experimentation in the wild around the CHANGES files/tag, we figured generating the version string from the git tags is not viable: users can use the library with no git history, some projects imports the project and change the git history, use tags for another version tracking etc.

So I believe the simplest option is to revert my changes, and revert back to using the CHANGES file as the source for the `--version` flag.
However, as this version is already something with no real meaning, I believe we can loosen up the requirements around updating this file.

Here is what I'd propose then:
 - remove canary/release branches. They were used back when SPIRV-Tools was very unstable. We don't do hotfix anyway, so no need.
 - Keep the CHANGES file as partial source for version tracking.
 
 
* 1 maintainer updated the CHANGES file to have an "editorialized" CHANGELOG
   (take the git log, filter important commits, write a nice changelog)
   This is submitted as a normal PR (review + merge)
   From now on, all builds at head will show vYYYY.X in the `--version`. That's fine.
   What's important is the sha1 of the release.
* 1 maintainer create a new release candidate tag `vYYYY.X.rc1` at `main:head`
* all partners are notified of this release candidate. They should test it and report back results (~1 week window delay)
   (mostly through internal communication channel, as not everything can be public)
* The maintainer who pushed the rc1 tag is notified of a failure. The test is fixed at `main:head`.
* 1 maintainer pushes a new tag `vYYYY.X.rc2`
* we go for another round of partner testing
* the maintainer who pushed the `rc2` tag received an "all good"
* the maintainer pushes a new tag at `vYYYY.X.rc2`: `vYYYY.X`
* a github workflow triggers on this tag
    - parses the `CHANGES` file to generate the changelog message
    - publishes a new release using this branch.

**Comment by s-perron** [comment]
I believe this can be closed now.  @Keenuts Have we made all of the changes we want to do?

**Comment by Keenuts** [comment]
Oh yes, this is done now :)
